programs/ido-program/programs/ido-program/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+
+    #[account(adress = pool_fiat.mint)]
+    pub fiat_mint: Account<'info, Mint>,
+
+    #[account(mut, constraint = pool_fiat.owner == *pool_signer.key)]
+    pub pool_fiat: Account<'info, TokenAccount>,
+
+    #[account(mut)]
+    pub payer: Signer<'info>,
+
+    #[account(mut)]
+    pub creator_fiat: Account<'info, TokenAccount>,
+
+    pub token_program: Program<'info, Token>,
+}
+
+#[account]
+pub struct PoolAccount {
+    /// Authority of the Pool
+    pub pool_authority: Pubkey,
+
+    /// Mint of redeemable tokens (Intermediate tokens which will be exchanged for native tokens)
+    pub redeemable_mint: Pubkey,
+
+    /// Mint of project tokens
+    pub native_mint: Pubkey,
+
+    /// Mint of fiat tokens
+    pub fiat_mint: Pubkey,
+
+    /// Token Account of Pool associated with the project token mint
+    pub pool_native: Pubkey,
+
+    /// Token Account of Pool associated with fiat mint
+    pub pool_fiat: Pubkey,
+
+    /// Total number of native tokens being distributed
+    pub total_native_tokens: u64,
+
+    /// Unix timestamp for starting IDO
+    pub start_ido_ts: i64,
+
+    /// Unix timestamp for ending IDO
+    pub end_ido_ts: i64,
+
+    /// Unix timestamp for withdrawing Fiat from pool
+    pub withdraw_fiat_ts: i64,
+
+    /// Bump
+    pub bump: u8,
+}
+
+impl PoolAccount {
+    pub const LEN: usize = DISCRIMINATOR_LENGTH   // Discriminator Length
+        + PUBKEY_LENGTH                           // Pool Authority
+        + PUBKEY_LENGTH                           // Redeemable Mint
+        + PUBKEY_LENGTH                           // Fiat Mint
+        + PUBKEY_LENGTH                           // Pool Native Token Account
+        + PUBKEY_LENGTH                           // Native Mint
+        + PUBKEY_LENGTH                           // Pool fiat Token Account
+        + DATA_LENGTH_64                          // Total Native Token Amount
+        + DATA_LENGTH_64                          // Start IDO TS
+        + DATA_LENGTH_64                          // End IDO TS
+        + DATA_LENGTH_64                          // Withdraw Fiat TS
+        + DATA_LENGTH_8; // Bump
+}
+
+const DISCRIMINATOR_LENGTH: usize = 8;
+const PUBKEY_LENGTH: usize = 32;
+const DATA_LENGTH_64: usize = 8;
+const DATA_LENGTH_8: usize = 1;
+
+#[error_code]
+pub enum ErrorCode {
+    #[msg("Timestamps are not Sequential")]
+    NonSequentialTimestamps,
+    #[msg("Invalid Parameter")]
+    InvalidParameter,
+    #[msg("Invalid Bump")]
+    InvalidBump,
+    #[msg("IDO has not begun yet")]
+    IdoFuture,
+    #[msg("Not the correct time to invest")]
+    WrongInvestingTime,
+    #[msg("Insufficient Fiat Tokens")]
+    LowFiat,
+    #[msg("IDO has not ended yet")]
+    IdoNotOver,
+    #[msg("Cannot withdraw Fiat yet")]
+    CannotWithdrawYet,
+}
+
+// Access Control Modifiers
+
+// IDO Starts in the Future
+fn pre_ido_phase<'info>(start_ido_ts: i64) -> Result<()> {
+    if !(Clock::get().unwrap().unix_timestamp < start_ido_ts) {
```

Comment:
`Clock::get().unwrap().unix_timestamp` is being multiple times, better to refactor it into it's own function.

```rs
pub fn get_timestamp() -> UnixTimestamp {
    Clock::get().unwrap().unix_timestamp
}
```